### PR TITLE
feat(android15) : Support Android 15 SDK-3997

### PR DESCRIPTION
**Pull request by piyush-kukadiya**
1. Update dependencies in [gradle/libs.versions.toml](https://github.com/CleverTap/clevertap-android-sdk/pull/657/files#diff-697f70cdd88ba88fe77eebda60c7e143f6ad1286bca75017421e93ad84fb87df)
2. Support Edge-to-Edge and display cutouts for Inbox
3. Support Edge-to-Edge and display cutouts for InApps (Footer, header, cover, cover-image,Footer-html, header-html, cover-html)
4. Fix `CTInAppWebview` `onMeasure()` calculations
5. Replace `onBackPressed()` with ` OnBackPressedCallback` for full screen click handling of in-app video 
6. Fix OpenJDKAPI changes 
7. Replace `ExistingPeriodicWorkPolicy.REPLACE` with `ExistingPeriodicWorkPolicy.UPDATE`
8. Remove `setPendingIntentBackgroundActivityStartMode()` usage
9. Fix build error for test shared module
10. Add tests for `String.format()`

### CODE REVIEW

File: clevertap-core/src/test/java/com/clevertap/android/sdk/CTXtensionsTest.kt [hunk 1]
```diff
@@ -1047,6 +1050,16 @@ class CTXtensionsTest : BaseTestCase() {
         assertFalse(result)
     }
 
+    @Test
+    fun `string format1`(){
```
Comment (vasct) [inline clevertap-core/src/test/java/com/clevertap/android/sdk/CTXtensionsTest.kt:1054]: Those two tests seem redundant

File: sample/src/main/java/com/clevertap/demo/MyApplication.kt [hunk 2]
```diff
@@ -147,6 +146,12 @@ class MyApplication : MultiDexApplication(), CTPushNotificationListener, Activit
                 //.penaltyDeath()
                 .build()
         )*/
+   /*     StrictMode.setVmPolicy(
```
Comment (vasct) [inline sample/src/main/java/com/clevertap/demo/MyApplication.kt:149]: Do we need this commented-out code ?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeInterstitialFragment.java [hunk 3]
```diff
@@ -61,6 +63,16 @@
     private FrameLayout videoFrameInDialog;
 
     private ViewGroup.LayoutParams imageViewLayoutParams;
+    private final OnBackPressedCallback onBackPressedCallback = new OnBackPressedCallback(false) {
+        @Override
+        public void handleOnBackPressed() {
+            if (exoPlayerFullscreen) {
+                config.getLogger().debug("Frag", "handleOnBackPressed() called");
```
Comment (vasct) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeInterstitialFragment.java:70]: A more descriptive tag to this log could be more useful

File: clevertap-core/src/test/java/com/clevertap/android/sdk/CTXtensionsTest.kt (hunk 1 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/CTXtensionsTest.kt:1054]: It's meant for testing `String.format()` usage in our SDK. `String.format()` has some breaking changes in Android 15.

File: sample/src/main/java/com/clevertap/demo/MyApplication.kt (hunk 2 above)
Comment (piyush-kukadiya) [reply sample/src/main/java/com/clevertap/demo/MyApplication.kt:149]: Yes, it's a toggle for testing purposes.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeInterstitialFragment.java (hunk 3 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeInterstitialFragment.java]: fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/dc0283ab69ceb9e8f9593826d016f8decb076280

File: clevertap-core/src/test/java/com/clevertap/android/sdk/CTXtensionsTest.kt (hunk 1 above)
Comment (vasct) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/CTXtensionsTest.kt:1054]: It seems a bit strange to me to test the SDK methods. Maybe we should test this with the methods where it is used in our code and verify that they still work correctly or wrap the String.format in our own function and test that. If this is the only way to test this, maybe at least compare the result of .format() with assertEquals to the expected behavior.

File: sample/src/main/java/com/clevertap/demo/MyApplication.kt (hunk 2 above)
Comment (vasct) [reply sample/src/main/java/com/clevertap/demo/MyApplication.kt:149]: Ok, maybe a comment above the block explaining what it's for would be useful.

File: clevertap-core/src/test/java/com/clevertap/android/sdk/CTXtensionsTest.kt (hunk 1 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/CTXtensionsTest.kt]: fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/5b7c660be6474a5204cf93ba2e6f97a86869ebf4

File: sample/src/main/java/com/clevertap/demo/MyApplication.kt (hunk 2 above)
Comment (piyush-kukadiya) [reply sample/src/main/java/com/clevertap/demo/MyApplication.kt]: Fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/c301c4cd97fefe36a6e4e6e8ef3c84e7c1f4baf9

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt [hunk 4]
```diff
@@ -299,3 +305,17 @@ fun String?.toJsonOrNull(): JSONObject? {
         }
     }
 }
+
+fun View.applySystemBarsInsetsWithMargin(block : (bars:Insets, mlp:MarginLayoutParams) -> Unit) {
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt:309]: why is there a need to create this extension function, seems to make the code unreadable where used.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt [hunk 5]
```diff
@@ -299,3 +305,17 @@ fun String?.toJsonOrNull(): JSONObject? {
...
+    ViewCompat.setOnApplyWindowInsetsListener(this
+    ) { v, insets ->
+        val bars: Insets = insets.getInsets(
+            WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
+        )
+        v.updateLayoutParams<MarginLayoutParams> {
+            block(bars,this)
+        }
+        WindowInsetsCompat.CONSUMED
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt:318]: why are we consuming the insets, if out work is done we can let it flow to view heirarchy.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt [hunk 6]
```diff
@@ -299,3 +305,17 @@ fun String?.toJsonOrNull(): JSONObject? {
         }
     }
 }
+
+fun View.applySystemBarsInsetsWithMargin(block : (bars:Insets, mlp:MarginLayoutParams) -> Unit) {
+    ViewCompat.setOnApplyWindowInsetsListener(this
+    ) { v, insets ->
+        val bars: Insets = insets.getInsets(
+            WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt:313]: System bars are fine, not sure what is `displayCutout()` , please make sure this is needed.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeHeaderFragment.java [hunk 7]
```diff
@@ -87,7 +92,12 @@ public boolean onTouch(View v, MotionEvent event) {
                 return true;
             }
         });
-
+        applySystemBarsInsetsWithMargin(inAppView, (insets, mlp) -> {
+            mlp.leftMargin = insets.left;
+            mlp.rightMargin = insets.right;
+            mlp.topMargin = insets.top;
+            return null;
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeHeaderFragment.java:99]: intentional skipping of `bottomMargin`?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppHtmlFooterFragment.java [hunk 8]
```diff
@@ -14,6 +19,13 @@ ViewGroup getLayout(View view) {
 
     @Override
     View getView(LayoutInflater inflater, ViewGroup container) {
-        return inflater.inflate(R.layout.inapp_html_footer, container, false);
+        View inAppView =  inflater.inflate(R.layout.inapp_html_footer, container, false);
+        applySystemBarsInsetsWithMargin(inAppView, (insets, mlp) -> {
+            mlp.leftMargin = insets.left;
+            mlp.rightMargin = insets.right;
+            mlp.bottomMargin = insets.bottom;
```
Comment (Anush-Shand) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppHtmlFooterFragment.java:26]: topMargin is required here?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppHtmlHeaderFragment.java [hunk 9]
```diff
@@ -14,7 +17,13 @@ ViewGroup getLayout(View view) {
 
     @Override
     View getView(LayoutInflater inflater, ViewGroup container) {
-        return inflater.inflate(R.layout.inapp_html_header, container, false);
+        View inAppView = inflater.inflate(R.layout.inapp_html_header, container, false);
+        applySystemBarsInsetsWithMargin(inAppView, (insets, mlp) -> {
+            mlp.leftMargin = insets.left;
+            mlp.rightMargin = insets.right;
+            mlp.topMargin = insets.top;
```
Comment (Anush-Shand) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppHtmlHeaderFragment.java:24]: bottomMargin is omitted here on purpose?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt [hunk 10]
```diff
@@ -299,3 +305,17 @@ fun String?.toJsonOrNull(): JSONObject? {
         }
     }
 }
+
+fun View.applySystemBarsInsetsWithMargin(block : (bars:Insets, mlp:MarginLayoutParams) -> Unit) {
+    ViewCompat.setOnApplyWindowInsetsListener(this
+    ) { v, insets ->
+        val bars: Insets = insets.getInsets(
```
Comment (Anush-Shand) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt:312]: Can we change this lambda to something with more readability?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppWebView.java [hunk 11]
```diff
@@ -47,7 +47,7 @@ public boolean performClick() {
     protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
         super.onMeasure(widthMeasureSpec, heightMeasureSpec);
         updateDimension();
-        setMeasuredDimension(dim.x, dim.y);
+        setMeasuredDimension(measureWidth(widthMeasureSpec,dim.x), measureHeight(heightMeasureSpec,dim.y));
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppWebView.java:50]: can you point out why this is changed?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inbox/CTInboxListViewFragment.java [hunk 12]
```diff
@@ -155,12 +159,24 @@ public void run() {
...
         return allView;
     }
 
+    private void applySystemBarsInsets(RecyclerView view) {
+        view.setClipToPadding(false);
+        ViewCompat.setOnApplyWindowInsetsListener(view, (v, insets) -> {
+            Insets bars = insets.getInsets(
+                    WindowInsetsCompat.Type.systemBars()
+            );
+            v.setPadding(bars.left, 0, bars.right, bars.bottom);
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inbox/CTInboxListViewFragment.java:175]: we are using paddings where instead of margins, also why do we not do this on parent and do it on RV?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/INotificationRenderer.java [hunk 13]
```diff
@@ -149,13 +149,8 @@ default NotificationCompat.Builder setActionButtons(
...
                                 actionLaunchIntent, flagsActionLaunchPendingIntent);
                     } else {
-                        Bundle optionsBundle = null;
-                        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
-                            optionsBundle = ActivityOptions.makeBasic().setPendingIntentBackgroundActivityStartMode(
-                                    ActivityOptions.MODE_BACKGROUND_ACTIVITY_START_ALLOWED).toBundle();
-                        }
                         actionIntent = PendingIntent.getActivity(context, requestCode,
-                                actionLaunchIntent, flagsActionLaunchPendingIntent, optionsBundle);
+                                actionLaunchIntent, flagsActionLaunchPendingIntent, null);
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/INotificationRenderer.java:153]: which android change does this pertain to?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java [hunk 14]
```diff
@@ -530,7 +530,7 @@ private void createOrResetWorker(boolean isPingFrequencyUpdated) {
 
                 String workName = existingWorkName.equals("") ? config.getAccountId() : existingWorkName;
 
-                workManager.enqueueUniquePeriodicWork(workName, ExistingPeriodicWorkPolicy.REPLACE, request);
+                workManager.enqueueUniquePeriodicWork(workName, ExistingPeriodicWorkPolicy.UPDATE, request);
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:533]: why are we changing the logic here?

File: clevertap-core/src/main/res/layout/inbox_activity.xml [hunk 15]
```diff
@@ -9,7 +9,8 @@
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
         android:elevation="4dp"
-        android:minHeight="56dp" />
+        android:minHeight="56dp"
+        android:fitsSystemWindows="true"/>
```
Comment (CTLalit) [inline clevertap-core/src/main/res/layout/inbox_activity.xml:13]: if we use this then what is the need to set insets (in form of paddings) to RV?

File: instantapp/build.gradle.kts [hunk 16]
```diff
@@ -1,16 +1,17 @@
 plugins {
     id("com.android.application")
     id("org.jetbrains.kotlin.android")
+    alias(libs.plugins.compose.compiler)
 }
 
 android {
     namespace = "com.clevertap.demo"
-    compileSdk = 33
+    compileSdk = 34
```
Comment (CTLalit) [inline instantapp/build.gradle.kts:9]: lets migrate to 35?

File: sample/src/main/AndroidManifest.xml [hunk 17]
```diff
@@ -15,7 +16,12 @@
         android:label="@string/app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
-        android:theme="@style/AppTheme">
+        android:theme="@style/AppTheme"
+        >
+        <property
+            android:name="android.adservices.AD_SERVICES_CONFIG"
+            android:resource="@xml/gma_ad_services_config"
+            tools:replace="android:resource" />
```
Comment (CTLalit) [inline sample/src/main/AndroidManifest.xml:24]: why is this added?

File: sample/src/main/AndroidManifest.xml [hunk 18]
```diff
@@ -119,7 +125,8 @@
         <activity
             android:name=".WebViewActivity"
             android:exported="true"
-            android:label="WebView">
+            android:label="WebView"
+            android:theme="@style/AppTheme.NoActionBar">
```
Comment (CTLalit) [inline sample/src/main/AndroidManifest.xml:129]: reason to add no action bar theme?

File: sample/src/main/java/com/clevertap/demo/MyApplication.kt [hunk 19]
```diff
@@ -132,8 +132,22 @@ class MyApplication : MultiDexApplication(), CTPushNotificationListener, Activit
...
+     *
+     * - For VM policy, it monitors potential memory-related issues such as leaks or misuse of API calls.
+     *   Violations are logged, helping developers find and fix memory management problems.
+     *
+     * This method is typically used during the development phase to catch bugs early. The `penaltyLog()`
+     * method ensures that violations are logged without crashing the app, but more aggressive actions
+     * like crashing can be enabled by using `penaltyDeath()` if desired.
+     */
     private fun setupStrictMode() {
-        /*StrictMode.setThreadPolicy(
```
Comment (CTLalit) [inline sample/src/main/java/com/clevertap/demo/MyApplication.kt:136]: we can make a different flavour for this.

File: sample/src/main/java/com/clevertap/demo/ui/main/HomeScreenFragment.kt [hunk 20]
```diff
@@ -220,4 +226,13 @@ class HomeScreenFragment : Fragment() {
...
     }
+
+    private fun applySystemBarsInsets(view: ExpandableListView) {
+        view.clipToPadding = false
+        ViewCompat.setOnApplyWindowInsetsListener(view) { v: View, insets: WindowInsetsCompat ->
+            val bars: Insets = insets.getInsets(WindowInsetsCompat.Type.systemBars())
+            v.setPadding(bars.left, 0, bars.right, bars.bottom)
+            WindowInsetsCompat.CONSUMED
+        }
+    }
```
Comment (CTLalit) [inline sample/src/main/java/com/clevertap/demo/ui/main/HomeScreenFragment.kt:237]: paddings to list and clip to padding are used, other places set insets as view margins

**Review by CTLalit — COMMENTED** [review]
Some comments+doubts are added.

File: instantapp/build.gradle.kts (hunk 16 above)
Comment (Anush-Shand) [inline instantapp/build.gradle.kts:9]: Should this be 35 as well?

**Review by Anush-Shand — COMMENTED** [review]
Left few comments

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt (hunk 4 above)
Comment (Anush-Shand) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt:309]: Can we rename this `block` function parameter?

Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt]: fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/c98e97d0528714f07fbfeee9ea451ad99931c5c0

Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt]: Resolving as discussed.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt (hunk 5 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt:341]: To avoid performance overhead and as given in official docs https://developer.android.com/develop/ui/views/layout/edge-to-edge#cutout-insets

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt (hunk 6 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/CTXtensions.kt:336]: To make UI compatible with top and bottom notches(displayCutout)

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeFooterFragment.java [hunk 21]
```diff
@@ -90,7 +94,13 @@ public boolean onTouch(View v, MotionEvent event) {
                 return true;
             }
         });
-
+        applySystemBarsInsetsWithMargin(inAppView, (insets, mlp) -> {
+            mlp.leftMargin = insets.left;
+            mlp.rightMargin = insets.right;
+            mlp.bottomMargin = insets.bottom;
+            return null;
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeFooterFragment.java]: Footer inApp don't require topMargin

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeHeaderFragment.java (hunk 7 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNativeHeaderFragment.java]: Header inApp don't require bottomMargin

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppWebView.java (hunk 11 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppWebView.java:50]: It's a bug when App is in edge-to-edge mode. Child must respect parent's required dimensions otherwise UI will break.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inbox/CTInboxListViewFragment.java (hunk 12 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inbox/CTInboxListViewFragment.java:175]: Resolving as discussed.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java (hunk 14 above)
Comment (CTLalit) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:533]: due to new api, deprecated "replace" in workmanager.

File: sample/src/main/AndroidManifest.xml (hunk 17 above)
Comment (CTLalit) [reply sample/src/main/AndroidManifest.xml:24]: discussed as per new google ads sdk we have to add this to avoid build errors.

**Review by CTLalit — APPROVED** [review]

**Review by Anush-Shand — APPROVED** [review]

**Comment by CTLalit** [comment]
Opened new PR for the same.
